src/GovUk.Education.ExploreEducationStatistics.Data.Services/FilterAndIndicatorViewModelBuilders.cs

```diff
@@ -0,0 +1,296 @@
+                labelSelector: value => value.Label,
+                sequenceIdSelector: sequenceEntry => sequenceEntry,
+                resultSelector: value => new IndicatorMetaViewModel(value),
+                sequence
+            );
+        }
+    }
+
+    private static IComparer<string> LabelComparer { get; } = new LabelRelationalComparer();
+
+    private static bool IsLabelTotal<T>(T input, Func<T, string> labelSelector)
+    {
+        return labelSelector(input).Equals("Total", StringComparison.InvariantCultureIgnoreCase);
+    }
+
+    private static IEnumerable<Ordered<TValue, TSequence>> OrderByLabel<TValue, TSequence>(
+        IEnumerable<TValue> values,
+        Func<TValue, string> labelSelector)
+    {
+        return values.OrderBy(labelSelector, LabelComparer)
+            .Select(source => new Ordered<TValue, TSequence>(source));
+    }
+
+    private static List<Ordered<TValue, TSequence>> OrderByLabelWithTotalFirst<TValue, TSequence>(
+        IEnumerable<TValue> values,
+        Func<TValue, string> labelSelector)
+    {
+        return values.OrderBy(value => !IsLabelTotal(value, labelSelector))
+            .ThenBy(labelSelector, LabelComparer)
+            .Select(source => new Ordered<TValue, TSequence>(source))
+            .ToList();
+    }
+
+    private static IEnumerable<Ordered<TValue, TSequence>> OrderBySequence<TValue, TSequence, TId>(
+        IEnumerable<TValue> values,
+        Func<TValue, TId> idSelector,
+        Func<TSequence, TId> sequenceIdSelector,
+        IEnumerable<TSequence> sequences) where TId : notnull
+    {
+        // TODO EES-1238 improve this to use the source list as the iteration base and throw exception if no ordering?
+        var valueMap = values.ToDictionary(idSelector);
+        foreach (var sequence in sequences)
+        {
+            var id = sequenceIdSelector(sequence);
+            if (valueMap.ContainsKey(id))
+            {
+                yield return new Ordered<TValue, TSequence>(valueMap[id], sequence);
+            }
+        }
+    }
+
+    private static Dictionary<string, TResult> OrderAsDictionary<TValue, TId, TSequence, TResult>(
+        IEnumerable<TValue> values,
+        Func<TValue, TId> idSelector,
+        Func<TValue, string> labelSelector,
+        Func<TSequence, TId> sequenceIdSelector,
+        Func<Ordered<TValue, TSequence>, int, TResult> resultSelector,
+        IEnumerable<TSequence>? sequence) where TId : notnull
+    {
+        var ordered = (sequence == null
+            ? OrderByLabel<TValue, TSequence>(values, labelSelector)
+            : OrderBySequence(values, idSelector, sequenceIdSelector, sequence)).ToList();
+
+        var result = new Dictionary<string, TResult>(ordered.Count);
+        ordered.ForEach((ordered1, index) =>
+        {
+            var key = labelSelector(ordered1).PascalCase();
+            var value = resultSelector(ordered1, index);
+            result.Add(key, value);
+        });
```

Comment:
Perhaps it might be convenient to create a `GroupByIndexed` enumerable extension method instead of manually creating this? Same applies for the grouping in `OrderAsDictionaryWithTotalFirst` below.

Reply:
Thanks that does simplify things a bit. I've made this change now.

Reply:
Update: just renamed this to `ToDictionaryIndexed` as it doesn't actually do a GroupBy. It works similar to the built in `ToDictionary` in the sense that if multiple values evaluate to the same key it will throw an error, rather than grouping the values into a list.